interpreter/beam/beam.go

```diff
@@ -33,29 +36,83 @@ var (
 )
 
 type beamData struct {
-	otpRelease         string
-	ertsVersion        string
-	theActiveCodeIndex libpf.Address
-	r                  libpf.Address
-	beamNormalExit     libpf.Address
-	ertsFrameLayout    uint64
-
+	otpRelease          string
+	ertsVersion         string
+	theActiveCodeIndex  libpf.Address
+	r                   libpf.Address
+	beamNormalExit      libpf.Address
+	ertsFrameLayout     uint64
+	ertsAtomTable       uint64
+	etpPtrMask          uint64
+	etpHeaderSubtagMask uint64
+	etpHeapBitsSubtag   uint64
 	// Sizes and offsets BEAM internal structs we need to traverse
 	vmStructs struct {
 		// ranges
 		// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/beam_ranges.c#L56-L61
 		ranges struct {
 			sizeOf uint8
 		}
+
+		// BeamCodeHeader
+		// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/beam_code.h#L56-L125
+		beamCodeHeader struct {
+			sizeOf, numFunctions, lineTable, functions uint8
+		}
+
+		// ErtsCodeInfo
+		// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/code_ix.h#L104-L123
+		ertsCodeInfo struct {
+			sizeOf, mfa uint8
+		}
+
+		// ErtsCodeMFA
+		// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/code_ix.h#L87-L95
+		ertsCodeMfa struct {
+			sizeOf, module, function, arity uint8
+		}
+
+		// BeamCodeLineTab
+		// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/beam_code.h#L130-L138
+		beamCodeLineTab struct {
+			sizeOf, fnamePtr, locSize, locTab, funcTab uint8
+		}
+
+		// IndexTable
+		// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/index.h#L39-L47
+		indexTable struct {
+			segTable uint8
+		}
+
+		// Atom
+		// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/atom.h#L48-L54
+		// In OTP 28, we need to look it up as a binary:
+		// https://github.com/erlang/otp/blob/OTP-28.0.2/erts/emulator/beam/atom.h#L50-L59
+		atom struct {
+			len, name uint8
+			u         struct {
+				bin uint8
+			}
+		}
+
+		// ErlHeapBits
+		// https://github.com/erlang/otp/blob/OTP-28.0.2/erts/emulator/beam/erl_bits.h#L149-L154
+		erlHeapBits struct {
+			data uint8
+		}
 	}
 }
 
 type beamInstance struct {
 	interpreter.InstanceStubs
 
-	pid  libpf.PID
-	data *beamData
-	rm   remotememory.RemoteMemory
+	pid         libpf.PID
+	data        *beamData
+	rm          remotememory.RemoteMemory
+	rangesPtr   libpf.Address
+	atomTable   libpf.Address
+	atomCache   map[uint32]string
```

Comment:
This is unbounded map. Would `LRU[uint32, libpf.String]` make more sense here? I think it would make sense to cap the cache size, and use `libpf.String` to keep things interned.